private/src/gst/gstavbvideosrc.cpp

```diff
@@ -0,0 +1,246 @@
+            ias_avbvideobridge_destroy_receiver(avbvideosrc->receiver);
+            avbvideosrc->receiver = NULL;
+            return GST_STATE_CHANGE_FAILURE;
+        }
+    }
+
+    ret = GST_ELEMENT_CLASS(gst_avbvideosrc_parent_class)->change_state(element, transition);
+    if (ret == GST_STATE_CHANGE_FAILURE) {
+        return ret;
+    }
+
+    GST_ERROR_OBJECT(avbvideosrc, "Parent state: %d [%d]", ret, GST_STATE_CHANGE_NO_PREROLL);
+
+    if (transition == GST_STATE_CHANGE_PLAYING_TO_PAUSED) {
+        ret = GST_STATE_CHANGE_NO_PREROLL;
+        avbvideosrc->done = TRUE;
+    }
+
+    if (transition == GST_STATE_CHANGE_READY_TO_NULL) {
+        ias_avbvideobridge_destroy_receiver(avbvideosrc->receiver);
+        avbvideosrc->receiver = NULL;
+    }
+
+    return ret;
+}
+
+static void
+gst_avbvideosrc_init(GstAvbVideoSrc *avbvideosrc)
+{
+    gst_base_src_set_live(GST_BASE_SRC(avbvideosrc), TRUE);
+    gst_base_src_set_format(GST_BASE_SRC(avbvideosrc), GST_FORMAT_TIME);
+
+    avbvideosrc->queue = g_async_queue_new();
+    if (!avbvideosrc->queue) {
+        GST_ERROR_OBJECT(avbvideosrc, "Could not create async queue");
+        return;
+    }
+
+    GST_OBJECT_FLAG_SET(avbvideosrc, GST_ELEMENT_FLAG_PROVIDE_CLOCK);
+
+    avbvideosrc->stream_name = strdup(DEFAULT_STREAM_NAME);
+    avbvideosrc->done = FALSE;
+}
+
+void
+gst_avbvideosrc_finalize(GObject * object)
+{
+    GstAvbVideoSrc *avbvideosrc = GST_AVBVIDEOSRC(object);
+
+    // TODO use g_async_queue_unref_full to clean up any remaining items
+    g_async_queue_unref(avbvideosrc->queue);
+    avbvideosrc->queue = NULL;
+
+    free(avbvideosrc->stream_name);
+    avbvideosrc->stream_name = NULL;
+
+    G_OBJECT_CLASS(gst_avbvideosrc_parent_class)->finalize(object);
+}
+
+/* ask the subclass to create a buffer with offset and size, the default
+ * implementation will call alloc and fill. */
+static GstFlowReturn
+gst_avbvideosrc_create(GstPushSrc * src, GstBuffer ** buf)
+{
+    GstAvbVideoSrc *avbvideosrc = GST_AVBVIDEOSRC(src);
+
+    do {
+        *buf = (GstBuffer *)g_async_queue_timeout_pop(avbvideosrc->queue, 100000);
+    } while (*buf == NULL && !avbvideosrc->done);
+
+    return GST_FLOW_OK;
+}
+
+static GstClock *
+gst_avbvideosrc_provide_clock(GstElement *element)
+{
+    (void)element;
+
+    return gst_avbclock_obtain();
+}
+
+static void
+mpegts_callback(ias_avbvideobridge_receiver *receiver, bool sph, ias_avbvideobridge_buffer const * packet, void *user_ptr)
+{
+    GstAvbVideoSrc *avbvideosrc = GST_AVBVIDEOSRC(user_ptr);
+    unsigned int expected_size = sph ? 192 : 188;
+    unsigned int offset = sph ? 4 : 0;
+    (void)receiver;
+
+    if ((packet->size % expected_size) != 0) {
+        GST_WARNING_OBJECT(avbvideosrc, "Packet may contain incomplete data. Size %zd is not a multiple of %u", packet->size, expected_size);
+    }
+
+    for (unsigned j = 0; packet->size - j >= expected_size; j += expected_size) {
```

Comment:
I think the stop condition of the `for` is wrong, as you are dealing with `unsigned int`s and `packet->size` is a `size_t` (also unsigned), you may not detect the overflow case. Does that make sense?

Reply:
You mean,  `j` overflowing before consuming all `packet->size`? That would be a problem sure. But then I'm at a loss to understand if the problem is `j` not being `size_t` or something else at the stop condition...

Reply:
For example, in that (error) case when `packet->size` is not a multiple of `expected_size`: `packet->size - j` will get promoted to an unsigned quantity, which will make the loop to iterate quite a few times more than you are expecting.

The problem I am seeing is that you won't be able to detect when `packet->size -j` becomes negative.

Reply:
Why would it become negative?
`packet->size` is constant, `j` increases by `expected_size` each iteration. The first loop in which `packet->size -j` is smaller than `expected_size` will end the loop (if it is a multiple, the number will be 0, if not, it will be some `x` such as `0 < x < expected_size`). So I don't see it becoming negative. 
The only issue here is `j` not being `size_t`, so it could overflow (highly unlikely, but better safe than sorry =D ).
Of course, if you can show me a counterexample, I'll see new things.

Reply:
What I was thinking was the case that `packet->size` is smaller than 188. But reading the code, the code is safe.

Sorry for the noise :-)